path/path.go

```diff
@@ -3,188 +3,301 @@ package path
 
 import (
 	"fmt"
-	"path"
+	gopath "path"
 	"strings"
 
-	cid "github.com/ipfs/go-cid"
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p/core/peer"
 )
 
-// A Path represents an ipfs content path:
-//   - <cid>/path/to/file
-//   - /ipfs/<cid>
-//   - /ipns/<cid>/path/to/folder
-//   - etc
-type Path string
-
-// ^^^
-// TODO: debate making this a private struct wrapped in a public interface
-// would allow us to control creation, and cache segments.
+const (
+	IPFSNamespace = "ipfs"
+	IPNSNamespace = "ipns"
+	IPLDNamespace = "ipld"
+)
 
-// FromString safely converts a string type to a Path type.
-func FromString(s string) Path {
-	return Path(s)
+// Path is a generic path. Paths must be prefixed with a valid prefix.
+type Path interface {
+	// String returns the path as a string.
+	String() string
+
+	// Namespace returns the first component of the path. For example, the namespace
+	// of "/ipfs/bafy" is "ipfs".
+	Namespace() string
+
+	// Mutable returns false if the data pointed to by this path is guaranteed to not
+	// change. Note that resolved mutable paths can be immutable.
+	Mutable() bool
+
+	// Root returns the [cid.Cid] of the root object of the path. Root can return
+	// [cid.Undef] for Mutable IPNS paths that use [DNSLink].
+	//
+	// [DNSLink]: https://dnslink.dev/
+	Root() cid.Cid
+
+	// Segments returns the different elements ofd a path, which are delimited
+	// by a forward slash ("/"). The leading slash must be ignored, that is, no
+	// segment should be empty.
+	Segments() []string
 }
 
-// FromCid safely converts a cid.Cid type to a Path type.
-func FromCid(c cid.Cid) Path {
-	return Path("/ipfs/" + c.String())
+// ResolvedPath is a [Path] which was resolved to the last resolvable node.
+type ResolvedPath interface {
+	Path
+
+	// Cid returns the [cid.Cid] of the node referenced by the path.
+	Cid() cid.Cid
+
+	// Remainder returns the unresolved parts of the path.
+	Remainder() string
 }
 
-// Segments returns the different elements of a path
-// (elements are delimited by a /).
-func (p Path) Segments() []string {
-	cleaned := path.Clean(string(p))
-	segments := strings.Split(cleaned, "/")
+// ImmutablePath is a [Path] which is guaranteed to return "false" to [Mutable].
+type ImmutablePath struct {
+	Path
+}
 
-	// Ignore leading slash
-	if len(segments[0]) == 0 {
-		segments = segments[1:]
+func NewImmutablePath(p Path) (ImmutablePath, error) {
+	if p.Mutable() {
+		return ImmutablePath{}, fmt.Errorf("path was expected to be immutable: %s", p.String())
 	}
 
-	return segments
+	return ImmutablePath{p}, nil
+}
+
+type path struct {
+	str       string
+	root      cid.Cid
+	namespace string
+}
+
+func (p *path) String() string {
```

Comment:
```suggestion
func (p path) String() string {
```
This is never mutated so no need to use a pointer.